src/refiners/foundationals/latent_diffusion/ella_adapter.py

```diff
@@ -0,0 +1,294 @@
+                    fl.UseContext("ella", "timestep_embedding"),
+                    fl.SiLU(),
+                    fl.Linear(time_embedding_dim, embedding_dim * 2, device=device, dtype=dtype),
+                ),
+            ),
+            fl.Lambda(self._scale_shift_tensors),
+        )
+
+        self._init_parameters()
+
+    def _init_parameters(self) -> None:
+        _linear: fl.Linear = self.find(fl.Linear)
+        nn.init.zeros_(_linear.weight)
+        nn.init.zeros_(_linear.bias)
+
+    def _scale_shift_tensors(self, x: Tensor, time_embedding: Tensor) -> Tensor:
+        shift, scale = time_embedding.chunk(2, dim=-1)
+        return x * (1 + scale) + shift
+
+
+class ParameterInitialized(fl.Parameter):
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        nn.init.normal_(self.weight, mean=0, std=self.dims[1] ** 0.5)
+
+
+class Latents(fl.Chain):
+    def __init__(
+        self,
+        num_latents: int,
+        width: int,
+        device: Device | str | None = None,
+        dtype: DType | None = None,
+    ) -> None:
+        super().__init__(
+            ParameterInitialized(
+                num_latents,
+                width,
+                device=device,
+                dtype=dtype,
+            ),
+        )
+
+
+class LatentsAdaLayerNorm(AdaLayerNorm):
+    pass
+
+
+class InputAdaLayerNorm(AdaLayerNorm):
+    pass
+
+
+class OutputAdaLayerNorm(AdaLayerNorm):
+    pass
+
+
+class _PerceiverAttentionBlock(fl.Chain):
+    def __init__(
+        self,
+        width: int,
+        num_heads: int,
+        timestep_embedding_dim: int,
+        device: Device | str | None = None,
+        dtype: DType | None = None,
+    ) -> None:
+        super().__init__(
+            fl.Sum(
+                fl.Parallel(
+                    fl.GetArg(0),
+                    fl.Chain(
+                        fl.Distribute(
+                            LatentsAdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
+                            InputAdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
+                        ),
+                        fl.Parallel(
+                            fl.GetArg(0),
+                            fl.Concatenate(fl.GetArg(0), fl.GetArg(1), dim=1),
+                            fl.Concatenate(fl.GetArg(0), fl.GetArg(1), dim=1),
+                        ),
+                        fl.Attention(embedding_dim=width, num_heads=num_heads, device=device, dtype=dtype),
+                    ),
+                )
+            ),
+            fl.Residual(
+                fl.Chain(
+                    OutputAdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
+                    fl.Linear(width, width * 4, device=device, dtype=dtype),
+                    SquaredReLU(),
+                    fl.Linear(width * 4, width, device=device, dtype=dtype),
+                )
+            ),
+        )
```

Comment:
There are quite a few things that need to be corrected here. fl.Sum(fl.Parallel(*args)) is the same as fl.Sum(*args). This is because when the input of a sum is a tuple, it will sum it, but you can also pass multiple layers. Also  fl.Chain(fl.Sum(*args1), fl.Residual(*args2)) is the same thing that fl.Sum(*(args1 + args2)). So you can simplify everything and use a fl.Sum. It should look like this (haven't tested it):
```
class _PerceiverAttentionBlock(fl.Sum):
    def __init__(
        self,
        width: int,
        num_heads: int,
        timestep_embedding_dim: int,
        device: Device | str | None = None,
        dtype: DType | None = None,
    ) -> None:
        super().__init__(
            fl.GetArg(0),
            fl.Chain(
                fl.Distribute(
                    LatentsAdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
                    InputAdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
                ),
                fl.Parallel(
                    fl.GetArg(0),
                    fl.Concatenate(fl.GetArg(0), fl.GetArg(1), dim=1),
                    fl.Concatenate(fl.GetArg(0), fl.GetArg(1), dim=1),
                ),
                fl.Attention(embedding_dim=width, num_heads=num_heads, device=device, dtype=dtype),
            ),
            fl.Chain(
                OutputAdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
                fl.Linear(width, width * 4, device=device, dtype=dtype),
                SquaredReLU(),
                fl.Linear(width * 4, width, device=device, dtype=dtype),
            ),
        )
  ```
  
Also, you should refrain from using GetArg too much. I think you should put the variable GetArg(1) in the context and remove all GetArg(0).

Reply:
You are right passing one of the inputs as a context will make the code readable and avoid using GetArg(). I ended up with the following:
`
class PerceiverAttentionBlock(fl.Chain):
    def __init__(
        self,
        width: int,
        num_heads: int,
        timestep_embedding_dim: int,
        device: Device | str | None = None,
        dtype: DType | None = None,
    ) -> None:
        super().__init__(
            fl.Passthrough(
                fl.UseContext("ella_perceiverresampler", "text_input"),
                InputAdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
                fl.SetContext("ella_perceiverresampler", "proj_text_input"),
            ),
            fl.Residual(
                LatentsAdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
                fl.Parallel(
                    fl.Identity(),
                    fl.Concatenate(fl.Identity(), fl.UseContext("ella_perceiverresampler", "proj_text_input"), dim=1),
                    fl.Concatenate(fl.Identity(), fl.UseContext("ella_perceiverresampler", "proj_text_input"), dim=1),
                ),
                fl.Attention(embedding_dim=width, num_heads=num_heads, device=device, dtype=dtype)
            ),
            fl.Residual(
                    OutputAdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
                    fl.Linear(width, width * 4, device=device, dtype=dtype),
                    SquaredReLU(),
                    fl.Linear(width * 4, width, device=device, dtype=dtype),
            ),
        )
`
what do you think ?

Reply:
the context "text_input" is set at the beginning of PerceiverResampler

Reply:
following Cedric comments, ill refactor the implementation using usecontext() and following the perceiverresampler implementation on ip-adapter.